address.md

```diff
@@ -0,0 +1,162 @@
+    Checksum(a Addressi) []byte
+    validateChecksum(data, cksm []byte) bool
+}
+```
+
+An exampple implementation of this interface is shown below:
+
+```go
+type Address struct {
+    // 0: main-net
+    // 1: test-net
+    network byte
+    
+    // 0: SECP256K1 Public Key
+    // 1: ActorID
+    // 2: Builtin Actor
+    // 3: BLS Public Key
+    typ byte
+    
+    // raw bytes containing the data associated with typ
+    data []byte
+}
+
+// checksum size
+const CKSM_LEN = 4
+
+// Network Byte
+const (
+    Mainnet = 0
+	Testnet = 1
+)
+
+// Type Byte
+const (
+    SECP256K1 = 0
+    ID = 1
+    Actor = 2
+    BLS = 3
+)
+
+// String Prefix
+const (
+    MAINNET_PREFIX = "fc"
+    TESTNET_PREFIX = "tf"
+)
+
+// New returns an Address for the specified network and type containing data and fails otherwise.
+func New(n byte, t byte, d []byte) Address {
+	if n != Mainnet && n != Testnet {
+        Fatal(ErrUnknownNetwork)
+	}
+	if t != SECP256K1 && t != ID && t != Actor && t != BLS {
+        Fatal(ErrUnknownType)
+	}
+    return Address{
+        network: n,
+        typ: t,
+        data: d,
+    }
+}
+
+// Marshal an Address to bytes
+func Marshal(a Address) []byte {
+    if a.typ == ID {
+        return []byte{a.network, a.typ, a.data}
+    }
+    return []byte{a.network, a.typ, a.data, Checksum(a)}
+}
+
+// Unmarshal bytes to an Address
+func Unmarshal(a []byte) Address {
+    if a[1] == ID {
+        return New(a[0], a[1], a[2:])
+    }
+    
+    raw := []byte{a[1], a[2:]}
+    data := raw[1 : len(raw) - CKSM_LEN]
+    cksm := raw[len(raw) - CKSM_LEN : len(raw)]
+    
+    if !validChecksum(data, cksm) {
+        Fatal(ErrInvalidCksm)
+    }
+    
+    return New(a[0], a[1], data)
+}
+
+// EncodeString encodes an Address to a string
+func EncodeString(a Address) string {
+	var prefix string
+    switch a.network {
+    case Mainnet:
+        prefix = MAINNET_PREFIX
+    case Testnet:
+        prefix = TESTNET_PREFIX
+    default:
+        return "<INVALID ADDRESS>"
+    }
+    
+    suffix := a.typ + a.data
+    if a.typ == ID {
+        return prefix + base58.Encode(suffix)
+    }
+    return prefix + base58.Encode(suffix + Checksum(suffix))
+}
+
+// DecodeString decoes a string to an Address
+func DecodeString(a string) Address {
+    var ntwk byte
+    switch a[:2] {
+    case MAINNET_PREFIX:
+        ntwk = Mainnet
+    case TESTNET_PREFIX:
+    	ntwk = Testnet
+    default:
+        Fatal(ErrUnknownNetwork)
+    }
+    
+    raw := base58.Decode(a[2:])
+    if raw[0] != ID {
+        return Address{
+            network: ntwk,
+            typ: raw[0],
+            data: raw[1:]
+        }
+    }
+    
+    if !validChecksum(raw) {
+        Fatal(ErrInvalidChecksum)
+    }
+    
+    return Address{
+        network: ntwk,
+        typ: raw[0],
+        data: raw[1: len(raw)- CKSM_LEN]
+    }
+}
+
+// Checksum is the last CKSM_LEN bytes the sha256 of Address Type and its data.
+func Checksum(a Address) [CKSM_LEN]byte {
```

Comment:
why are we using sha256 suddenly as checksums? we had a proper scheme with error detections in the previous spec version

Reply:
What scheme were we previously using? I assume you mean [this one](https://github.com/filecoin-project/go-filecoin/blob/efdc601098fec6d71f1543327090c1e468d0836b/address/address.go#L327-L373), if so we continue to use it -- `sha256` was chosen here for simplicity in writing as I wasn't sure on the details of the aforementioned code.

Reply:
@dignifiedquire the previous scheme was just a complicated checksum, its far simpler to use a few bytes from a well known hash function.

Reply:
Based on other discussion we have had we are going with the first 4 bytes of the blake2b hash of an address